sdk/couchbase/src/mutation_state.rs

```diff
@@ -0,0 +1,253 @@
+use couchbase_core::queryx::query_options::{ScanVectorEntry, SparseScanVectors};
+use serde::de::{MapAccess, Visitor};
+use serde::ser::{SerializeMap, SerializeStruct};
+use serde::{de, Deserialize, Deserializer, Serialize, Serializer};
+use std::collections::HashMap;
+use std::fmt;
+use std::fmt::Write;
+
+#[derive(Clone, Debug, Eq, PartialEq, Ord, PartialOrd, Hash)]
+pub struct MutationToken {
+    token: couchbase_core::mutationtoken::MutationToken,
+    bucket_name: String,
+}
+
+impl MutationToken {
+    pub(crate) fn new(
+        token: couchbase_core::mutationtoken::MutationToken,
+        bucket_name: String,
+    ) -> Self {
+        Self { token, bucket_name }
+    }
+}
+
+#[derive(Default, Clone, Debug, Eq, PartialEq)]
+pub struct MutationState {
+    // There's a bit of duplication between key and value here but that allows us to recreate
+    // the tokens easily.
+    tokens: HashMap<MutationStateKey, couchbase_core::mutationtoken::MutationToken>,
+}
+
+#[derive(Clone, Debug, Eq, PartialEq, Ord, PartialOrd, Hash)]
+struct MutationStateKey {
+    bucket_name: String,
+    vbid: u16,
+}
+
+impl Serialize for MutationStateKey {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        let mut map = serializer.serialize_map(Some(1))?;
+        map.serialize_entry(&self.bucket_name, &self.vbid)?;
+        map.end()
+    }
+}
+
+impl MutationState {
+    pub fn new() -> Self {
+        Self {
+            tokens: HashMap::default(),
+        }
+    }
+
+    pub fn new_with_tokens(tokens: Vec<MutationToken>) -> Self {
+        let mut state = Self {
+            tokens: HashMap::default(),
+        };
+        for token in tokens {
+            let key = MutationStateKey {
+                bucket_name: token.bucket_name,
+                vbid: token.token.vbid,
+            };
+            if let Some(entry) = state.tokens.get(&key) {
+                if entry.seqno < token.token.seqno {
+                    state.tokens.insert(key, token.token);
+                }
+            } else {
+                state.tokens.insert(key, token.token);
+            }
+        }
+
+        state
+    }
+
+    pub fn push_token(&mut self, token: MutationToken) {
+        let key = MutationStateKey {
+            bucket_name: token.bucket_name,
+            vbid: token.token.vbid,
+        };
+
+        if let Some(entry) = self.tokens.get(&key) {
+            if entry.seqno < token.token.seqno {
+                self.tokens.insert(key, token.token);
+            }
+        } else {
+            self.tokens.insert(key, token.token);
+        }
+    }
+
+    pub fn tokens(&self) -> Vec<MutationToken> {
+        self.tokens
+            .iter()
+            .map(|(key, token)| MutationToken::new(token.clone(), key.bucket_name.clone()))
```

Comment:
Why use the bucket name from the key and not the token?

Reply:
There is no bucket name in the token. The token is the core representation rather than the one in this file.